src/interpolations.jl

```diff
@@ -1330,33 +1362,58 @@ function getPermLagrangeQ(order)
     return result
 end
 
-function ArbitraryOrderLagrange{RefQuadrilateral, order}(points::Vector{Float64} = GaussQuadrature.legendre(order+1, GaussQuadrature.both)[1]) where order
-    product_of = ArbitraryOrderLagrange{RefLine,order}(points)
-    ref_coord = Array{Vec{2,Float64},1}(undef,(order+1)^2)
-    for i in 1:order+1, j in 1:order+1
-        ref_coord[i+(j-1)*(order+1)] = Vec(points[i],points[j])
-    end
-    perm = getPermLagrangeQ(order)
-    inv_perm = sortperm(perm)
-    ArbitraryOrderLagrange{RefQuadrilateral, order, typeof(product_of), typeof(ref_coord), typeof(perm)}(product_of, ref_coord, perm, inv_perm)
-end
 
-getnbasefunctions(::ArbitraryOrderLagrange{RefQuadrilateral,order}) where order = (order + 1)^2
-
-function facedof_interior_indices(::ArbitraryOrderLagrange{RefQuadrilateral,order}) where order
-    return (SVector{order-1}((i+4 for i in 1:order-1)),
-        SVector{order-1}((i+order+3 for i in 1:order-1)),
-        SVector{order-1}((i+2*order+2 for i in 1:order-1)),
-        SVector{order-1}((i+3*order+1 for i in 1:order-1)))
+for (name,  default_coords) in (
+    (:ArbitraryOrderLagrange,               :both),
+    (:ArbitraryOrderDiscontinuousLagrange,  :neither)
+    )
+    @eval begin
+        function $(name){RefQuadrilateral, order}(points::Vector{Float64} = GaussQuadrature.legendre(order+1, GaussQuadrature.$(default_coords))[1]) where order
+            product_of = $(name){RefLine,order}(points)
+            ref_coord = Array{Vec{2,Float64},1}(undef,(order+1)^2)
+            for i in 1:order+1, j in 1:order+1
+                ref_coord[i+(j-1)*(order+1)] = Vec(points[i],points[j])
+            end
+            perm = getPermLagrangeQ(order)
+            inv_perm = sortperm(perm)
+            $(name){RefQuadrilateral, order, typeof(product_of), typeof(ref_coord), typeof(perm)}(product_of, ref_coord, perm, inv_perm)
+        end
+        getnbasefunctions(::$(name){RefQuadrilateral,order}) where order = (order + 1)^2
+    end
 end
 
-function facedof_indices(ip::ArbitraryOrderLagrange{RefQuadrilateral,order}) where order
-    interior = facedof_interior_indices(ip)
-    face1 = SVector{order + 1}((i == 1 ? 1 : i == 2 ? 2 : interior[1][i-2] for i in 1:order+1))    
-    face2 = SVector{order + 1}((i == 1 ? 2 : i == 2 ? 3 : interior[2][i-2] for i in 1:order+1))    
-    face3 = SVector{order + 1}((i == 1 ? 3 : i == 2 ? 4 : interior[3][i-2] for i in 1:order+1))    
-    face4 = SVector{order + 1}((i == 1 ? 4 : i == 2 ? 1 : interior[4][i-2] for i in 1:order+1))    
-    return (face1, face2, face3, face4)
+for (name,  facefuncs) in (
+    (:ArbitraryOrderLagrange,               :facedof_indices),
+    (:ArbitraryOrderDiscontinuousLagrange,  :dirichlet_facedof_indices)
```

Comment:
I just realized that using GL points should result in `dirichlet_facedof_indices` being empty IIUC. This makes this function need to either check for faces on boundary or have predetermined options for the points (equidistant, GL, Radau) instead of passing the points in the constructor.
I think having predetermined options would be easier and more consistent (as we guarantee the ordering), what do you think?

Reply:
I would suggest that we just throw an error for now when trying to extract the Dirichlet dofs for all L2 elements which are not based on "symmetric" nodes, when the basis functions of the interior nodes are non-zero on the boundary (and for modal basis functions). If we want to work further on strong enforcement of the Dirichlet condition we have to think a bit how to do it correctly.